### PR TITLE
logins: API to rekey the database

### LINKED ISSUE #1988: [logins] API to rekey the database
There's a good chance Fenix will ship logins storage with keys that we'd want to change once secure storage story is figured out, and we'd need to rekey the database file. This issue is about exposing a method that will call `pragma rekey=newkey`.

https://www.zetetic.net/sqlcipher/sqlcipher-api/#Changing_Key

┆Issue is synchronized with this [Jira Story](https://jira.mozilla.com/browse/SYNC-53)
┆Story Points: 3
┆Epic: Logins Component 1.0
┆Sprint: SYNC - end 2019-12-20

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #1988

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt [hunk 1]
```diff
@@ -207,6 +207,13 @@ class DatabaseLoginsStorage(private val dbPath: String) : AutoCloseable, LoginsS
         }
     }
 
+    @Throws(LoginsStorageException::class)
+    override fun rekeyDatabase(newEncryptionKey: String) {
```
Comment (thomcc) [inline components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt]: We ended up having to add a byte-array version of the other API (which hex encodes). Do we need to do that for this too?

Comment (vladikoff) [reply components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt]: Added, also rebased the PR

File: components/logins/ffi/src/lib.rs [hunk 2]
```diff
@@ -98,6 +98,42 @@ pub extern "C" fn sync15_passwords_disable_mem_security(handle: u64, error: &mut
...
+/// Dereferences the `new_encryption_key` pointer, and is thus unsafe.
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_rekey_database_with_hex_key(
+    handle: u64,
+    new_encryption_key: *const u8,
+    new_encryption_key_len: u32,
+    error: &mut ExternError,
+) {
+    log::debug!("sync15_passwords_rekey_database_with_hex_key");
+    let new_key = bytes_to_key_string(new_encryption_key, new_encryption_key_len as usize).unwrap();
```
Comment (thomcc) [inline components/logins/ffi/src/lib.rs]: This should happen inside `ENGINES.call_with_result`.

File: components/logins/src/db.rs [hunk 3]
```diff
@@ -90,6 +90,16 @@ impl LoginDb {
         Ok(())
     }
 
+    // Change the key on an existing encrypted database,
```
Comment (thomcc) [inline components/logins/src/db.rs]: Nit: `///` for doc comments

**Pull request merged**